keps/sig-node/3322-add-maxRestartTimes-to-podSpec/README.md

```diff
@@ -0,0 +1,973 @@
+The "Design Details" section below is for the real
+nitty-gritty.
+-->
+
+Add a new field `maxRestartTimesOnFailure` to the `pod.spec`. `maxRestartTimesOnFailure` only applies
+when `restartPolicy` is set to `OnFailure`. As a result, the restart times of a Pod
+will be controlled by `maxRestartTimesOnFailure`.
+
+### User Stories (Optional)
+
+<!--
+Detail the things that people will be able to do if this KEP is implemented.
+Include as much detail as possible so that people can understand the "how" of
+the system. The goal here is to make this feel real for users without getting
+bogged down.
+-->
+
+#### Story 1
+
+When running with a pod, some system errors or program internal errors
+encountered, which causes the application restarting infinitely. This is
+unrecoverable and wasteful.
+
+### Story 2
+
+For batch Job specifically, now a Job provides a single backoffLimit for all derived pods,
+which may cause the entire Job failed and clean up all the resources because of one of the pods
+restarts for more than the `spec.backoffLimit`. To solve the problem, the Batch WG incubates
+a new KEP [Backoff Limit Per Index For Indexed Jobs](https://github.com/kubernetes/enhancements/issues/3850)
+which introduces a new field `backoffLimitPerIndex` to control the max restart times per Pod.
+
+It's actually what we want to implement here, by sinking the capacity of `backoffLimitPerIndex` to pod level,
+it helps to reduce the maintenance costs of the Job API.
+
+Note: currently, `backoffLimitPerIndex` only supports `restartPolicy=Never` initially, but as the KEP states, it will depend
+on this feature to support onFailure restartPolicy in the future, see [chapter](https://github.com/kubernetes/enhancements/blob/ab54c14de86af3bcc76b45a4c9194bc47f6c9a62/keps/sig-apps/3850-backoff-limits-per-index-for-indexed-jobs/README.md#support-backofflimitperindex-when-restartpolicyonfailure).
+This reminds me whether we should widen the _max_restart_times_ to other restart policies like `Never`, see discussion below.
+
+### Notes/Constraints/Caveats (Optional)
+
+<!--
+What are the caveats to the proposal?
+What are some important details that didn't come across above?
+Go in to as much detail as necessary here.
+This might be a good place to talk about core concepts and how they relate.
+-->
+
+#### Discussions
+
+The Batch WG incubates a new KEP [Backoff Limit Per Index For Indexed Jobs](https://github.com/kubernetes/enhancements/issues/3850),
+which introduces a new field `backoffLimitPerIndex`, then each pod of the Job will have an independent backOffLimit,
+but initially it only supports `restartPolicy=Never`, maybe we can widen the _max_restart_times_ to support other restart policies,
+like `Never`, then the Job API no longer needs to consider these logics anymore. Currently, it uses an annotation `batch.kubernetes.io/job-index-failure-count`
+to record the retires number, which looks not elegant.
+
+Pros:
+  * Reduce the maintenance cost of Job API
+
+Cons:
```

Comment:
I think a more important con, in the context of Jobs, is that a pod-level API can't keep a counter across pod recreations, which will happen when a node fails, or needs to be updated, or common scenarios like eviction and preemption

Reply:
Why would we want to track this across recreations?  At that point it's a new pod

Reply:
But it's still the same Job. Let's say as an administrator, I want to allow up to 5 restarts (otherwise I consider the job buggy). If a pod fails 3 times, then gets preempted, the new pod should only be allowed to fail 2 times.

Reply:
Which brings a good use case for this API in combination with KEP-3850:
The user only sets the Job level limit, then the job controller can create the pod with less retries every time the pod is disrupted.

cc @mimowo

Reply:
Job controller, when recreating a pod with restartPolicy=OnFailure could create the new one with the job-failure-count=3 annotation. 

To do this, Job controller needs to delay finalizer removal until the pod is recreated (similarly as in KEP-3850).

In case we have an indexed job, then job controller could delay finalizer removal based on the presence of `job.spec.template.spec.maxRestartTimes`, but it does not need to require `backoffLimitPerIndex` - it could work for any indexed job.

Certainly, we should coordinate to reuse the code for delayed finalizer removal, and trigger it in this new scenario.

Reply:
> But it's still the same Job.

But it's not the same pod.  Job is a higher-level contruct.  Nothing about one pod intriniscally links it to some other "previous attempt" pod.  We don't have that concept (for better or worse).  If you want to thread two pods together at a hiugher level there are primitives to use, but you havr to orchestrate them from the higher-order controller.

Reply:
I know, but the "Motivation" section of this KEP explicitly talks about batch Jobs, so it's fair to list this as a downside. Or to find ways to reconcile fields in both APIs.

Reply:
> But it's not the same pod. Job is a higher-level contruct. 

Yes, it feels like a conflation of a pod-level `maxRestarts` and a job-level `maxRecreates` or something.

Reply:
> Yes, it feels like a conflation of a pod-level `maxRestarts` and a job-level `maxRecreates` or something.

Or a job-level `maxRestarts` that creates new pods with `job.maxRestarts - previousPodMaxRestarts` ?